Build error from null unmanaged view in device code

**Describe the bug**

Constructing an unmanaged view in HIP device code works fine if a general pointer is provided, but if NULL/nullptr is provided there's a compilation error.

**Please include the following for a minimal reproducer**

- rocm 6.4.1 or 6.4.2 (probably other version too)
- Kokkos develop f6973de29
- HIP backend for arch AMD_GFX942_APU

Reproducer config:
```
cmake -DCMAKE_CXX_COMPILER=hipcc -DKokkos_ENABLE_HIP=ON -DKokkos_ARCH_AMD_GFX942_APU=ON -DCMAKE_BUILD_TYPE=RELEASE ../kokkos/example/build_cmake_in_tree
```
Reproducing program example/build_cmake_in_tree/cmake_example.cpp:
```
#include <Kokkos_Core.hpp>
#include <cstdio>
#include <iostream>

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc, argv);
  Kokkos::DefaultExecutionSpace{}.print_configuration(std::cout);
  {
    Kokkos::parallel_for(Kokkos::RangePolicy<>(0, 10),
    KOKKOS_LAMBDA(int i) {
	    Kokkos::View<int*, Kokkos::HIPSpace> myView(nullptr, 3);
	});
  }
  Kokkos::finalize();
  return 0;
}
```
Compiling this gives:
```
kokkos/example/build_cmake_in_tree/cmake_example.cpp:11:43: error: reference to __host__ function 'View<int>' in __host__ __device__ function
   11 |             Kokkos::View<int*, Kokkos::HIPSpace> myView(nullptr, 3);
      |                                                  ^
kokkos/core/src/HIP/Kokkos_HIP_ParallelFor_Range.hpp:44:5: note: called by 'exec_range<void>'
   44 |     m_functor(i);
      |     ^
kokkos/core/src/HIP/Kokkos_HIP_ParallelFor_Range.hpp:70:22: note: called by 'operator()'
   70 |       this->template exec_range<WorkTag>(iwork);
      |                      ^
kokkos/core/src/HIP/Kokkos_HIP_KernelLaunch.hpp:76:3: note: called by 'hip_parallel_launch_local_memory<Kokkos::Impl::ParallelFor<(lambda at kokkos/example/build_cmake_in_tree/cmake_example.cpp:10:5), Kokkos::RangePolicy<>>>'
   76 |   driver();
      |   ^
kokkos/core/src/Kokkos_View.hpp:734:12: note: 'View<int>' declared here
  734 |   explicit View(decltype(nullptr), Args... args)
      |            ^
1 error generated when compiling for gfx942.
```
Possibly related to #7427.

--- a/core/src/Kokkos_View.hpp
+++ b/core/src/Kokkos_View.hpp
@@ -670,12 +670,12 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
   // Special function to be preferred over the above for passing in 0, NULL or
   // nullptr when pointer type is char*
   template <class... Args>
-  explicit View(decltype(nullptr), Args... args)
+  KOKKOS_FUNCTION explicit View(decltype(nullptr), Args... args)
       : View(Kokkos::view_wrap(pointer_type(nullptr)), args...) {}
 
   // FIXME: Constructor which allows always 8 sizes should be deprecated
   template <class... P>
-  explicit View(
+  KOKKOS_FUNCTION explicit View(
       const Impl::ViewCtorProp<P...>& arg_prop,
       std::enable_if_t<!Impl::ViewCtorProp<P...>::has_pointer, const size_t>
           arg_N0          = KOKKOS_IMPL_CTOR_DEFAULT_ARG,

--- a/core/src/Kokkos_View.hpp
+++ b/core/src/Kokkos_View.hpp
@@ -675,7 +675,7 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
 
   // FIXME: Constructor which allows always 8 sizes should be deprecated
   template <class... P>
-  KOKKOS_FUNCTION explicit View(
+  explicit View(
       const Impl::ViewCtorProp<P...>& arg_prop,
       std::enable_if_t<!Impl::ViewCtorProp<P...>::has_pointer, const size_t>
           arg_N0          = KOKKOS_IMPL_CTOR_DEFAULT_ARG,

--- a/core/src/Kokkos_View.hpp
+++ b/core/src/Kokkos_View.hpp
@@ -670,7 +670,7 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
   // Special function to be preferred over the above for passing in 0, NULL or
   // nullptr when pointer type is char*
   template <class... Args>
-  KOKKOS_FUNCTION explicit View(decltype(nullptr), Args... args)
+  KOKKOS_FUNCTION explicit View(std::nullptr_t, Args... args)
       : View(Kokkos::view_wrap(pointer_type(nullptr)), args...) {}
 
   // FIXME: Constructor which allows always 8 sizes should be deprecated

--- a/core/unit_test/TestViewEmptyRuntimeUnmanaged.hpp
+++ b/core/unit_test/TestViewEmptyRuntimeUnmanaged.hpp
@@ -21,36 +21,44 @@
 namespace {
 
 template <class T>
-void test_empty_view_runtime_unmanaged() {
-  T d{};
+struct TestEmptyViewRuntimeUnmanaged {
+  template <class ExecutionSpace>
+  TestEmptyViewRuntimeUnmanaged(ExecutionSpace const& exec) {
+    Kokkos::parallel_for(Kokkos::RangePolicy(exec, 0, 1), *this);
+  }
+
+  KOKKOS_FUNCTION void operator()(int) const {
+    T d{};
 #if defined(_MSC_VER)
 #pragma warning(push)
 #pragma warning(disable : 4312)
 #endif
-  auto* p = reinterpret_cast<T*>(0xABADBABE);
+    auto* p = reinterpret_cast<T*>(0xABADBABE);
 #if defined(_MSC_VER)
 #pragma warning(push)
 #endif
 
-  (void)Kokkos::View<T*>(p, 0);
-  (void)Kokkos::View<T*>(&d, 0);
-  (void)Kokkos::View<T*>(nullptr, 0);
-  (void)Kokkos::View<T*>(NULL, 0);  // NOLINT(modernize-use-nullptr)
-  (void)Kokkos::View<T*>(0, 0);     // NOLINT(modernize-use-nullptr)
+    (void)Kokkos::View<T*>(p, 0);
+    (void)Kokkos::View<T*>(&d, 0);
+    (void)Kokkos::View<T*>(nullptr, 0);
+    (void)Kokkos::View<T*>(NULL, 0);  // NOLINT(modernize-use-nullptr)
+    (void)Kokkos::View<T*>(0, 0);     // NOLINT(modernize-use-nullptr)
 
-  (void)Kokkos::View<T**>(p, 0, 0);
-  (void)Kokkos::View<T**>(&d, 0, 0);
-  (void)Kokkos::View<T**>(nullptr, 0, 0);
-  (void)Kokkos::View<T**>(NULL, 0, 0);  // NOLINT(modernize-use-nullptr)
-  (void)Kokkos::View<T**>(0, 0, 0);     // NOLINT(modernize-use-nullptr)
-}
+    (void)Kokkos::View<T**>(p, 0, 0);
+    (void)Kokkos::View<T**>(&d, 0, 0);
+    (void)Kokkos::View<T**>(nullptr, 0, 0);
+    (void)Kokkos::View<T**>(NULL, 0, 0);  // NOLINT(modernize-use-nullptr)
+    (void)Kokkos::View<T**>(0, 0, 0);     // NOLINT(modernize-use-nullptr)
+  }
+};
 
 TEST(TEST_CATEGORY, view_empty_runtime_unmanaged) {
-  test_empty_view_runtime_unmanaged<float>();
-  test_empty_view_runtime_unmanaged<const double>();
-  test_empty_view_runtime_unmanaged<int>();
-  test_empty_view_runtime_unmanaged<char>();
-  test_empty_view_runtime_unmanaged<const char>();
+  TEST_EXECSPACE exec;
+  (void)TestEmptyViewRuntimeUnmanaged<float>(exec);
+  (void)TestEmptyViewRuntimeUnmanaged<const double>(exec);
+  (void)TestEmptyViewRuntimeUnmanaged<int>(exec);
+  (void)TestEmptyViewRuntimeUnmanaged<char>(exec);
+  (void)TestEmptyViewRuntimeUnmanaged<const char>(exec);
 }
 
 }  // namespace

